### ISSUE TITLE
WHO commands sent without awaiting previous response

### ISSUE THREAD

**Issue by hahuang65** [weechat/weechat#1726]
## Bug summary

Connecting to libera thru chat.sr.ht's bouncer (soju) causes a bunch of 

```
WHO :This command could not be completed because it has been used recently, and is rate-limited.
```
to occur.

After discussing with `@emersion` from `#sr.ht` and `@jess` from `#libera`, it was determined from my `/server raw` log that weechat is sending WHO commands every 4 seconds, without waiting for a reply from the previous WHO command.
They mentioned that there should not be simultaneous WHO commands being run.

This happens when (re-)connecting to the server and clears up after a few minutes (when the WHO commands finally resolve)

Attaching my logs [irc.irc_raw.weechatlog.log](https://github.com/weechat/weechat/files/7645041/irc.irc_raw.weechatlog.log)

- WeeChat version: 3.3
- OS, distribution and version: macOS 10.15.7
- Terminal: kitty
- Terminal multiplexer (screen/tmux/…/none): none

**Comment by flashcode** [comment]
Hi,

Did you change the value of option `irc.server_default.away_check` (which is 0 = disabled by default)?
Or the same option in a specific server?

If you enabled one of these options, please try to set value to 0 and tell me if it fixes the problem.
If yes, that means it is caused by the check of away nicks (which should be the case after a quick look in the code).

**Comment by hahuang65** [comment]
@flashcode nope, it's still set to 0. It's also set to 0 for the specific server that I'm on.

**Comment by flashcode** [comment]
Additional questions:

* Do you have scripts loaded? If yes, can you try without any script loaded?
* Does it happen when you connect directly to libera.chat, without the bouncer?
* Please give the exact steps to reproduce the problem, including how to configure WeeChat from the default config, so that I can try to reproduce on my side.

**Comment by emersion** [comment]
I think the 2s delay comes from `anti_flood_prio_low`.

Instead of relying on antiflood, weechat should avoid sending a new `WHO` request before it got the reply from the previous one.

**Comment by flashcode** [comment]
@emersion: yes probably, but I need to understand why:

* many WHO are sent are same time when `irc.server_default.away_check` is set to 0
* the problem does not happen when connecting directly to servers (at least I never saw such issue, even if I join a lot of channels)

**Comment by emersion** [comment]
The WHO messages come from: https://github.com/weechat/weechat/blob/f4a1baaed3921bc88b699a81a8c9a75acb5ac09d/src/plugins/irc/irc-channel.c#L774

WeeChat sends `WHO #channel` for each `JOIN` it receives (and the server supports various extensions), to get the initial away status and account names for members of the channel.

**Comment by flashcode** [comment]
Yes, but why don't I see any problem while connecting to servers?
This seems specific to the soju bouncer (which I didn't test).

**Comment by emersion** [comment]
If I had to guess why this issue happens more often with a bouncer:

- Bouncers multiplex connections from different clients. If I have 2 clients connected to my bouncer, and I wake them up together (e.g. resume my laptop, and a few seconds later, unlock my phone), then both clients will send `WHO` requests at the same time. Note, soju does have a `WHO` queue, and will not send the next `WHO` request before the previous one completed.
- Some servers give clients 1 `WHO` credit when they `JOIN` a new channel. Since the bouncer is always joined, it doesn't get this extra credit when its own clients connect and is subject to the regular `WHO` rate limiting.
- Bouncers add latency. So if the server sends a long reply, maybe it'll be fine with a direct connection, maybe it'll exceed 2s with a bouncer.

It doesn't seem like @hahuang65 was able to reproduce consistently this bug, sometimes it happens and sometimes it's fine. So it likely depends on whether the 2s interval is enough for the reply to come through or not, how many users in each channel there are, and so on.

In any case, the consensus in #ircdocs is that clients should wait before sending new `WHO` commands.

**Comment by flashcode** [comment]
Yes, I'll have to introduce some mechanism to prevent two WHO to be sent at same time, waiting for the previous response.
Do you have any link to the docs you're mentioning?
Thanks.

**Comment by emersion** [comment]
Sweet!

> Do you have any link to the docs you're mentioning?

Unfortunately no. When I wrote "#ircdocs", I meant the #ircdocs channel on Libera Chat, ie. the channel for https://modern.ircdocs.horse/. We discussed this issue there.
